io/pci/dlpar.py

```diff
@@ -0,0 +1,293 @@
+#!/usr/bin/python
+
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: 2017 IBM
+# Author: Pridhiviraj Paidipeddi <ppaidipe@linux.vnet.ibm.com>
+# Author: Venkat Rao B <vrbagal1@linux.vnet.ibm.com>
+
+import os
+import shutil
+try:
+    import pxssh
+except ImportError:
+    from pexpect import pxssh
+from avocado import Test
+from avocado import main
+from avocado.utils import process
+from avocado.utils import distro
+from avocado.utils import pci
+from avocado.utils.software_manager import SoftwareManager
+from avocado.utils.process import CmdError
+
+
+class CommandFailed(Exception):
+    def __init__(self, command, output, exitcode):
+        self.command = command
+        self.output = output
+        self.exitcode = exitcode
+
+    def __str__(self):
+        return "Command '%s' exited with %d.\nOutput:\n%s" \
+               % (self.command, self.exitcode, self.output)
+
+
+class DlparPci(Test):
+    '''
+    DLPAR PCI script does pci add,remove and also move operation from one
+    lpar to another lpar. Update the details in yaml file.
+    For move operation, please configure another lpar and update in yaml file.
+    And also make sure both rsct and rsct_rm services up and running
+    '''
+
+    def setUp(self):
+        '''
+        set up required packages and gather necessary test inputs
+        '''
+        sm = SoftwareManager()
+        detected_distro = distro.detect()
+        self.log.info("Test is running on:" + detected_distro.name)
+        if not sm.check_installed("ksh") and not sm.install("ksh"):
+            self.error('ksh is needed for the test to be run')
+        if detected_distro.name == "Ubuntu":
+            if not sm.check_installed("python-paramiko") and not \
+                                      sm.install("python-paramiko"):
+                self.error('python-paramiko is needed for the test to be run')
+            ubuntu_url = self.params.get('ubuntu_url', default=None)
+            debs = self.params.get('debs', default=None)
+            for deb in debs:
+                deb_url = os.path.join(ubuntu_url, deb)
+                deb_install = self.fetch_asset(deb_url, expire='7d')
+                shutil.copy(deb_install, self.srcdir)
+                process.system("dpkg -i %s/%s" % (self.srcdir, deb),
+                               ignore_status=True, sudo=True)
+        else:
+            url = 'http://ausgsa.ibm.com/projects/p/poweryum/ibmtools/ibmtools'
```

Comment:
was the internal link accessible outside ? what if the link does not work ? is the script expected to be available openly ?

Reply:
hmm...Good catch...I am not sure about the link. But w.r.t., script being openly available, yes. We have conformation from sachin.

Reply:
@abdhaleegit agreed. We can move the `url` parameter to yaml file with it being empty by default.
@vrbagalkote please.